pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -198,50 +198,54 @@ impl FileExt for EphemeralFile {
+                Writer {
+                    blknum: (ephemeral_file.size / PAGE_SZ as u64) as u32,
+                    off: (ephemeral_file.size % PAGE_SZ as u64) as usize,
+                    ephemeral_file,
+                }
+            }
+            fn push_bytes(&mut self, src: &[u8]) -> Result<(), io::Error> {
+                let mut src_remaining = src;
+                while !src_remaining.is_empty() {
+                    {
+                        let mut head_page = self.ephemeral_file.get_buf_for_write(self.blknum)?;
+                        let dst_remaining = &mut head_page[self.off..];
+                        let n = min(dst_remaining.len(), src_remaining.len());
+                        dst_remaining[..n].copy_from_slice(&src_remaining[..n]);
+                        self.off += n;
+                        src_remaining = &src_remaining[n..];
+                    }
+                    if self.off == PAGE_SZ {
+                        // This block is done, move to next one.
+                        self.blknum += 1;
+                        self.off = 0;
+                    }
+                }
+                Ok(())
+            }
+        }
 
-        let mut buf = self.get_buf_for_write(blknum)?;
+        let pos = self.size;
+        let mut writer = Writer::new(self);
 
         // Write the length field
         if srcbuf.len() < 0x80 {
-            buf[off] = srcbuf.len() as u8;
-            off += 1;
+            writer.push_bytes(&[srcbuf.len() as u8])?;
         } else {
             let mut len_buf = u32::to_be_bytes(srcbuf.len() as u32);
             len_buf[0] |= 0x80;
-            let thislen = PAGE_SZ - off;
-            if thislen < 4 {
-                // it needs to be split across pages
-                buf[off..(off + thislen)].copy_from_slice(&len_buf[..thislen]);
-                blknum += 1;
-                buf = self.get_buf_for_write(blknum)?;
-                buf[0..4 - thislen].copy_from_slice(&len_buf[thislen..]);
-                off = 4 - thislen;
-            } else {
-                buf[off..off + 4].copy_from_slice(&len_buf);
-                off += 4;
-            }
+            writer.push_bytes(&len_buf)?;
         }
 
         // Write the payload
-        let mut buf_remain = srcbuf;
-        while !buf_remain.is_empty() {
-            let mut page_remain = PAGE_SZ - off;
-            if page_remain == 0 {
-                blknum += 1;
-                buf = self.get_buf_for_write(blknum)?;
-                off = 0;
-                page_remain = PAGE_SZ;
-            }
-            let this_blk_len = min(page_remain, buf_remain.len());
-            buf[off..(off + this_blk_len)].copy_from_slice(&buf_remain[..this_blk_len]);
-            off += this_blk_len;
-            buf_remain = &buf_remain[this_blk_len..];
-        }
-        drop(buf);
+        writer.push_bytes(srcbuf)?;
 
         if srcbuf.len() < 0x80 {
```

Comment:
This wasn't introduced by this PR but it looks like that the size is updated only after all writes were successful. If only part of the writes went through, the size is not updated and it looks like the invariant that the rest of the code assumes, that size encodes the actually written bytes, is violated. I'm not sure how much of that is an issue, whether the size of ephemeral files is accessed after a write error, or the entire ephemeral file is discarded. But still, maybe one could integrate the size updating into `push_bytes`?

Reply:
The error handling and particularly retryability story was/remains quite 💩 

Here's my list of known problems:

* what happens if we retry after an error?
  * think about the scenario where write_blob needs to write 3 pages.
  * there are 3 opportunities for failure
  * if we fail and bail on the second opportunity, then we've written to the file on disk, and the page cache contains the changes, but, at least `size` isn't updated
  * a subsequent `read_blob` will return half-written data (the unwritten bytes will be zero-filled)
  * a subsequent `write_blob` will overwrite the written data

I was toying with the idea of bumping `size` in the `if self.off == PAGE_SZ`. 
But that makes the retry situation even worse, because now the subsequent `write_blob` doesn't overwrite, but instead it appends.

I propose to create an issue referencing this comment here, and fix it in a future PR.

Reply:
https://github.com/neondatabase/neon/issues/5007